Sources/WhoopDIKit/Module/AccumulationDefinition.swift

```diff
@@ -0,0 +1,121 @@
+        })
+    } else {
+        serviceDictionary[definition.serviceKey] = definition
+    }
+}
+
+
+/// Provides the definition of an accumulation that is recalculated on each request.
+/// A fresh accumulation will be computed each time it is requested.
+final class FactoryAccumulationDefinition: DependencyDefinition {
+    private let valueProvider: (Any?, Container) throws -> Any
+    private let insertionFunc: (ServiceDictionary<DependencyDefinition>, FactoryAccumulationDefinition) -> Void
+    let serviceKey: ServiceKey
+
+    init<Key: AccumulationKey>(name: String?, accumulationKey: Key.Type, valueProvider: @escaping (Any?) throws -> Key.AccumulatedValue) {
+        self.serviceKey = ServiceKey(Key.FinalValue.self, name: name)
+        self.valueProvider = wrapAccumulationProvider(name: name, accumulationKey: accumulationKey, valueProvider: valueProvider)
+        self.insertionFunc = { dictionary, definition in accumulate(key: accumulationKey,
+                                                                    definition: definition,
+                                                                    accumulatedValueGetter: valueProvider,
+                                                                    serviceDictionary: dictionary)
+        }
+    }
+
+    func get(params: Any?, container: Container) throws -> Any {
+        try valueProvider(params, container)
+    }
+
+    func insert(into serviceDictionary: ServiceDictionary<any DependencyDefinition>) {
+        insertionFunc(serviceDictionary, self)
+    }
+}
+
+/// Provides the definition of an accumulation that is calculated once and cached.
+/// The accumulated value is computed once per container and reused on subsequent requests.
+final class SingletonAccumulationDefinition: DependencyDefinition {
+    private let accumulationKeyType: Any.Type
+    private let valueProvider: (Any?, Container) throws -> Any
+    private let lock = NSLock()
+    let serviceKey: ServiceKey
+    private let insertionFunc: (ServiceDictionary<DependencyDefinition>, SingletonAccumulationDefinition) -> Void
+    private var cachedAccumulated: Any? = nil
+    private var cachedValue: Any? = nil
+
+    init<Key: AccumulationKey>(name: String?, accumulationKey: Key.Type, valueProvider: @escaping (Any?) throws -> Key.AccumulatedValue) {
+        self.accumulationKeyType = accumulationKey
+        self.serviceKey = ServiceKey(Key.FinalValue.self, name: name)
+        self.valueProvider = wrapAccumulationProvider(name: name, accumulationKey: accumulationKey, valueProvider: valueProvider)
+        self.insertionFunc = { dictionary, definition in accumulate(key: accumulationKey,
+                                                                    definition: definition,
+                                                                    accumulatedValueGetter: { params in
+            try definition.lockValue(externalGet: {
```

Comment:
The formatting there is super hard to read, I think we'd want to extract some of this out so it's lined up better in the editor.